### PR TITLE
Added rake task for finding broken links, and view to display them.

### PR DESCRIPTION
**Pull request by AndrewMcBurney**
## Overview
- Rake task can be scheduled to run nightly
- Adds new `/broken_links` view, to display broken links across articles

## Todo
- [x] A bit of design love on the `/broken_links` view

### CODE REVIEW

File: lib/tasks/audit_links.rake [hunk 1]
```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
+
+desc "Nigthly task to update list of broken article urls"
```
Comment (brupm) [inline lib/tasks/audit_links.rake]: I'd schedule this to run at 1am PST.

File: app/controllers/broken_links_controller.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+  # GET /broken_links
+  def index
+    @broken_links_array = broken_links_per_article
+  end
+
+  private
+
+  # Two dimensional array grouping articles containing broken links
+  def broken_links_per_article
+    BrokenLink.uniq.pluck(:article_id).map { |id| BrokenLink.where(article_id: id) }
```
Comment (brupm) [inline app/controllers/broken_links_controller.rb]: add a DB index to the table for article_id

File: app/helpers/html_with_pygments.rb [hunk 3]
```diff
@@ -59,7 +59,9 @@ def preprocess(full_document)
   private
 
   def article_status(link)
-    'article-not-found' if !valid_article?(link)
+    return if valid_article?(link)
+    BrokenLink.create!(url: link) if performing_link_audit?
+    "article-not-found"
```
Comment (brupm) [inline app/helpers/html_with_pygments.rb]: Another nice addition here would be to pull the logic of creating the BrokenLink into a new method, and upon regular article saving you can check if the link has been fixed and the remove it from the list, therefore updating the /broken_links path right away, so it doesn't give the person the impression there are still broken links until the next rake run.

Comment (AndrewMcBurney) [inline app/helpers/html_with_pygments.rb]: I'll create a factory module and include it where we want to create broken links. Does that sound okay?

File: app/controllers/articles_controller.rb [hunk 4]
```diff
@@ -117,6 +119,7 @@ def report_outdated
   def update
     if @article.update_attributes(article_params)
       flash[:notice] = "Article was successfully updated."
+      parse_article(@article, build_links: true)
```
Comment (AndrewMcBurney) [inline app/controllers/articles_controller.rb]: Rename to be more descriptive, add comment explaining what this method does.

File: app/models/broken_link.rb [hunk 5]
```diff
@@ -0,0 +1,6 @@
+# frozen_string_literal: true
+
+class BrokenLink < ApplicationRecord
+  belongs_to :article
+  belongs_to :author, class_name: "User"
```
Comment (AndrewMcBurney) [inline app/models/broken_link.rb]: Add `validates` for `url`

File: app/helpers/broken_links_helper.rb [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+  end
+
+  private
+
+  # Removes all old 'BrokenLink' records for a given article
+  def clear_links_for_article(article)
+    BrokenLink.where(article_id: article.id).delete_all
+  end
+
+  def article_options(article)
```
Comment (chriswoodrich) [inline app/helpers/broken_links_helper.rb]: this method abstraction doesn't really add value. best to let `find_broken_links_for_article`calculate its own arguments for the methods it calls

File: app/models/broken_link.rb [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+  validates  :article, :author, :url, presence: true
+
+  # Two dimensional array grouping articles containing broken links
+  def self.per_article
+    BrokenLink.uniq.pluck(:article_id)
```
Comment (chriswoodrich) [inline app/models/broken_link.rb]: you don't need to reference the class name from within a class method. `.uniq.pluck(:article_id)` should work itself (and won't break if you change the model name)

File: app/helpers/broken_links_helper.rb [hunk 8]
```diff
@@ -0,0 +1,27 @@
+# frozen_string_literal: true
```
Comment (chriswoodrich) [inline app/helpers/broken_links_helper.rb]: This isn't a great use case for a helper given the places that its `include`d are so disparate.  Better to do this a single responsibility class, with robust tests of the public api

File: spec/controllers/categories_controller_spec.rb [hunk 9]
```diff
@@ -5,6 +5,10 @@
 RSpec.describe CategoriesController, type: :controller do
   let(:attributes) { { label: "Sample Category" } }
 
+  before do
+    allow_any_instance_of(User).to(receive(:administrator?).and_return(true))
```
Comment (chriswoodrich) [inline spec/controllers/categories_controller_spec.rb]: is this needed on all specs?  If not, I would move it down to the context it's actually used.  It poses a small trap for the next person adding tests

File: spec/decorators/category_decorator_spec.rb [hunk 10]
```diff
@@ -1,6 +0,0 @@
-# frozen_string_literal: true
-
-require "rails_helper"
-
```
Comment (chriswoodrich) [inline spec/decorators/category_decorator_spec.rb]: 👍

File: spec/services/article_links_service_spec.rb [hunk 11]
```diff
@@ -0,0 +1,41 @@
+# frozen_string_literal: true
+
+require "rails_helper"
+
+RSpec.describe ArticleLinksService do
```
Comment (chriswoodrich) [inline spec/services/article_links_service_spec.rb]: nice

File: app/views/broken_links/_broken_links.haml [hunk 12]
```diff
@@ -0,0 +1,8 @@
+- @broken_links_array.each do |broken_links|
+  - unless broken_links.empty?
+    - article = broken_links.first.article
+
+    %li.list-item
+      %span.dib.mrxs
+      = link_to article.title, article
+      %strong - #{article.author.name} | #{broken_links.length}
```
Comment (chriswoodrich) [inline app/views/broken_links/_broken_links.haml]: lol public shaming

Comment (AndrewMcBurney) [inline app/views/broken_links/_broken_links.haml]: It was in the ticket ahahahahah 😂

File: spec/controllers/categories_controller_spec.rb (hunk 9 above)
Comment (AndrewMcBurney) [inline spec/controllers/categories_controller_spec.rb]: Ahh, good point 👍

**Pull request merged**